feat(ui): the toolbox became vertical

neSpecc opened the pull request:
This PR is the second part of the big UI update called "Unified Toolbox". See the [Roadmap](https://github.com/codex-team/editor.js#roadmap) for more details.

In this iteration, the Toolbox became a vertical list.

<img width="987" alt="image" src="https://user-images.githubusercontent.com/3684889/162225504-c65e9d3f-9228-4377-bc27-5d41a85b33de.png">

<img width="988" alt="image" src="https://user-images.githubusercontent.com/3684889/162225587-558056bd-96b1-41e5-bc8a-18a9d0a98d73.png">

## Changelog

- the `Popover` utility has been created. It will be used for all vertical lists.
- the `Toolbox` now uses the `Popover` to display the tools list
- the `Plus` button will be always shown (previously, it appears only for empty blocks) 

<img width="426" alt="image" src="https://user-images.githubusercontent.com/3684889/162226878-c8bca963-5a95-409f-b693-15c46a6b528c.png">

- the `Toolbar` won't move on hover at mobile viewports. Resolves #1972 
- the `Flipper` now scrolls to the focused item using `scrollIntoViewIfNeeded`


---

Also, see the Discussion https://github.com/codex-team/editor.js/discussions/1811

---

Thanks to @TatianaFomina for finishing the PR so fast.

neSpecc committed:
the popover component, vertical toolbox

neSpecc committed:
toolbox position improved

neSpecc committed:
popover width improved

neSpecc committed:
always show the plus button

neSpecc committed:
search field added

neSpecc committed:
Merge branch 'next' into feat/vertical-toolbox

neSpecc committed:
search input in popover

neSpecc committed:
trying to create mobile toolbox

neSpecc committed:
Merge branch 'next' into feat/vertical-toolbox

neSpecc committed:
Merge branch 'next' into feat/vertical-toolbox

TatianaFomina committed:
feat(toolbox): popover adapted for mobile devices (#2004)

* FIx mobile popover fixed positioning

* Add mobile popover overlay

* Hide mobile popover on scroll

* Alter toolbox buttons hover

* Fix closing popover on overlay click

* Tests fix

* Fix onchange test

* restore focus after toolbox closing by ESC

* don't move toolbar by block-hover on mobile

Resolves #1972

* popover mobile styles improved

* Cleanup

* Remove scroll event listener

* Lock scroll on mobile

* don't show shortcuts in mobile popover

* Change data attr name

* Remove unused styles

* Remove unused listeners

* disable hover on mobile popover

* Scroll fix

* Lint

* Revert "Scroll fix"

This reverts commit 82deae543eadd5c76b9466e7533bf3070d82ac4c.

* Return back background color for active state of toolbox buttons

Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

gohabereg commented on src/components/flipper.ts:
```diff
@@ -266,4 +266,15 @@ export default class Flipper {
+  private flipCallback(): void {
+    if (this.iterator.currentItem) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+      // @ts-ignore
```
Why ts-ignore here?

gohabereg commented on src/components/modules/toolbar/index.ts:
```diff
@@ -490,6 +509,15 @@ export default class Toolbar extends Module<ToolbarNodes> {
+       */
+      if (_.isMobile()) {
+        return;
+      }
```
Can we just do not subscribe on mobile?

TatianaFomina replied:
fixed

gohabereg commented on src/components/polyfills.ts:
```diff
@@ -96,3 +96,40 @@ if (!Element.prototype.prepend) {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
+  Element.prototype.scrollIntoViewIfNeeded = function (centerIfNeeded): void {
+    centerIfNeeded = arguments.length === 0 ? true : !!centerIfNeeded;
```
You can extends `Element` interface to avoid ts-ignore here

TatianaFomina replied:
fixed

gohabereg commented on src/components/ui/toolbox.ts:
```diff
@@ -124,67 +115,79 @@ export default class Toolbox extends EventsDispatcher<ToolboxEvent> {
 
-    this.addTools();
-    this.enableFlipper();
+    this.popover.on(PopoverEvent.OverlayClicked, () => {
```
Do we unsubscribe somewhere?

TatianaFomina replied:
fixed

gohabereg commented on src/components/utils.ts:
```diff
@@ -762,3 +762,10 @@ export function cacheable<Target, Value, Arguments extends unknown[] = unknown[]
+/**
+ * True if screen has mobile size
+ */
+export function isMobile(): boolean {
```
I suggest to rename the function as it doesn't return what expected. Something like `isNarrowScreen`

TatianaFomina replied:
renamed to `isMobileScreen`

gohabereg commented on src/components/utils/popover.ts:
```diff
@@ -0,0 +1,352 @@
+    documentScrollLocked: string;
+    } {
+    return {
+      popover: 'ce-popover',
+      popoverOpened: 'ce-popover--opened',
+      itemsWrapper: 'ce-popover__items',
+      item: 'ce-popover__item',
+      itemHidden: 'ce-popover__item--hidden',
+      itemFocused: 'ce-popover__item--focused',
+      itemLabel: 'ce-popover__item-label',
+      itemIcon: 'ce-popover__item-icon',
+      itemSecondaryLabel: 'ce-popover__item-secondary-label',
+      noFoundMessage: 'ce-popover__no-found',
+      noFoundMessageShown: 'ce-popover__no-found--shown',
+      popoverOverlay: 'ce-popover__overlay',
+      popoverOverlayHidden: 'ce-popover__overlay--hidden',
+      documentScrollLocked: 'ce-scroll-locked',
```
We should really think of using css modules

neSpecc replied:
We will, at some of the next updates

gohabereg commented on src/components/utils/popover.ts:
```diff
@@ -0,0 +1,352 @@
+  private createItem(item: PopoverItem): HTMLElement {
+    const el = Dom.make('div', Popover.CSS.item);
+
+    el.setAttribute('data-item-name', item.name);
```
```suggestion
    el.dataset.itemName =  item.name;
```

gohabereg commented on src/components/utils/search-input.ts:
```diff
@@ -0,0 +1,120 @@
+   */
+  private render(placeholder: string): void {
+    this.wrapper = Dom.make('div', SearchInput.CSS.wrapper);
+    const iconWrapper = Dom.make('div', SearchInput.CSS.icon);
```
```suggestion

    const iconWrapper = Dom.make('div', SearchInput.CSS.icon);
```

gohabereg commented on src/styles/input.css:
```diff
@@ -0,0 +1,46 @@
+img, video, audio, canvas, input,
```
Not encapsulated selectors

gohabereg replied:
Still relevant

TatianaFomina replied:
removed these lines

TatianaFomina committed:
Vertical toolbox fixes (#2017)

* Replace visibility property with display for hiding popover

* Disable arrow right and left keys for popover

* Revert "Replace visibility property with display for hiding popover"

This reverts commit af521cf6f29fb06b71a0e2e8ec88d6a757f9144f.

* Hide popover via setting max-height to 0 to fix animation in safari

* Remove redundant condition

TatianaFomina committed:
Extend element interface to avoid ts errors

TatianaFomina committed:
Do not subscribe to block hovered if mobile

TatianaFomina committed:
Add unsubscribing from overlay click event

TatianaFomina committed:
Rename isMobile to isMobileScreen

TatianaFomina committed:
Cleanup

gohabereg commented on src/components/flipper.ts:
```diff
@@ -206,23 +206,7 @@ export default class Flipper {
-    }
-
-    return this.activated && handlingKeyCodeList.indexOf(event.keyCode) !== -1;
+    return this.activated && this.allowedKeys.indexOf(event.keyCode) !== -1;
```
```suggestion
    return this.activated && this.allowedKeys.includes(event.keyCode);
```

gohabereg commented on src/components/utils/popover.ts:
```diff
@@ -0,0 +1,358 @@
+
+    if (item.secondaryLabel) {
+      el.appendChild(Dom.make('div', Popover.CSS.itemSecondaryLabel, {
+        innerHTML: item.secondaryLabel,
```
Do we really need `innerHTML` here? Can we use `textCintent`?

TatianaFomina committed:
fix: popover opening direction (#2022)

* Change popover opening direction based on available space below it

* Update check

* Use cacheable decorator

TatianaFomina committed:
Update src/components/flipper.ts

Co-authored-by: George Berezhnoy <gohabereg@users.noreply.github.com>

TatianaFomina committed:
Fixes

TatianaFomina committed:
Merge branch 'next' into feat/vertical-toolbox

TatianaFomina committed:
Merge branch 'feat/vertical-toolbox' of https://github.com/codex-team/editor.js into feat/vertical-toolbox

TatianaFomina committed:
Fix test

TatianaFomina committed:
Clear search on popover hide

TatianaFomina committed:
Fix popover width

TatianaFomina committed:
Fix for tests

TatianaFomina committed:
Update todos

TatianaFomina committed:
Linter fixes

neSpecc committed:
rm todo about beforeInsert

because I have no idea what does it mean

neSpecc committed:
i18n for search labels done

neSpecc committed:
rm methods for hiding/showing of +

neSpecc committed:
some code style update

neSpecc committed:
Merge branch 'next' into feat/vertical-toolbox

neSpecc committed:
Update CHANGELOG.md

neSpecc committed:
make the list items a little bit compact

neSpecc committed:
fix z-index issue caused by block-appearing animation

also, improve popover padding for two reasons:

- make the popover more consistent with the Table tool popover (in future, it can be done with the same api method)
- make popover looks better
